src/actions/attack.cpp

```diff
@@ -1644,13 +1644,15 @@ std::pair<int, bool> ability_leadership(const std::string& ability,const unit_ma
 	unit_ability_list abil = un->get_abilities(ability, weapon, opp_weapon);
 	for(unit_ability_list::iterator i = abil.begin(); i != abil.end();) {
 		const config &filter = (*i->first).child("filter_opponent");
+		const config &filter_attacker = (*i->first).child("filter_attacker");
+		const config &filter_defender = (*i->first).child("filter_defender");
 		bool show_result = false;
-		if(up == units.end() && !filter) {
+		if(up == units.end() && !filter && !filter_attacker && !filter_defender) {
 			show_result = un->abilities_filter_matches(*i->first, attacker, abil_value);
-		} else if(up == units.end() && filter) {
+		} else if(up == units.end() && (filter || filter_attacker || filter_defender)) {
 			return {abil_value, false};
 		} else {
-			show_result = !(!un->abilities_filter_matches(*i->first, attacker, abil_value) || !up->ability_filter_opponent(ability, *i->first, opp_loc));
+			show_result = !(!un->abilities_filter_matches(*i->first, attacker, abil_value) || !up->ability_filter_fighter(ability, "filter_opponent", *i->first, opp_loc) || ( attacker && !un->ability_filter_fighter(ability, "filter_attacker", *i->first, loc)) || ( !attacker && !up->ability_filter_fighter(ability, "filter_attacker", *i->first, opp_loc)) || (!attacker && !un->ability_filter_fighter(ability, "filter_defender", *i->first, loc)) || (attacker && !up->ability_filter_fighter(ability, "filter_defender", *i->first, opp_loc)));
```

Comment:
Can you split this up somehow so that the line is not ridiculously long? Since the exact same logic is used a second time in `bool_leadership`, maybe add a static function that dispatches the four calls to `ability_filter_fighter`.